### PR TITLE
【MIIT program】Add GMTNet for dielectric tensor prediction

### PR DESCRIPTION
**Pull request by cjywahaha-oss**
## Summary

This PR adds GMTNet support for dielectric tensor prediction in PaddleMaterials.

## Main changes

- Add the Paddle implementation of GMTNet under `ppmat/models/gmtnet/`
- Add GMTNet dataset, graph conversion, collate, and dataset preparation support
- Register GMTNet in the public model and dataset APIs
- Add canonical train/validation/test split metadata
- Integrate GMTNet with the public training and evaluation workflows
- Add `PropertyPredictor` support for:
  - Mapping inputs
  - Single and multiple `pymatgen.Structure` inputs
  - Single CIF files
  - CIF directories
- Add GMTNet-specific precision-preserving CIF parsing:
  - `primitive=False`
  - `sort=False`
  - `merge_tol=0.0`
  - `frac_tolerance=0.0`
- Ensure CIF directories are fully parsed before prediction starts, avoiding partial prediction when one file is invalid
- Add packaging support for GMTNet configuration and split resources
- Delay optional `paddle_scatter` imports for models that require it, so GMTNet can be imported and installed independently
- Remove the obsolete standalone GMTNet training script in favor of the public Trainer workflow

## Validation

The following validations were completed:

- Converted checkpoint strict loading
- Full public training workflow
- Validation and test evaluation for converted, best, and final checkpoints
- PropertyPredictor acceptance tests: 31/31 passed
- Mapping, Structure, CIF, and CIF-directory prediction
- Invalid input and checkpoint error handling
- CIF-directory atomic failure behavior
- Non-GMTNet CIF loader regression
- CPU-only inference with forward gradients enabled
- Wheel and sdist build
- Clean wheel and sdist installation from `site-packages`
- Package-content and sensitive-path audit

No checkpoint, training dataset, generated logs, or local `paddle_scatter` sources are included in the package.

## Known limitations

- A deterministic edge-ordering candidate was evaluated but not adopted because equivalent edge permutations caused widespread float32 reduction-order differences.
- A deterministic rotation canonicalization candidate was evaluated but not adopted because it changed the existing constraint baseline across the canonical test set.
- Dielectric tensors predicted from CIF inputs are expressed in the Cartesian frame of the parsed structure.
- Low-precision CIF files cannot recover floating-point information that was not encoded in the file.
- GMTNet evaluation and prediction require forward gradients to remain enabled because the model evaluates an analytical gradient block.

Related issue: #194

### CODE REVIEW

**Comment by leeleolay** [comment]
辛苦参考 https://github.com/PaddlePaddle/PaddleMaterials/issues/258 和套件内已有代码和实现的规范，并提交数据和权重 给出精度对齐结果

**Comment by leeleolay** [comment]
参考已有规范：https://github.com/PaddlePaddle/PaddleMaterials/issues/258

**Comment by cjywahaha-oss** [comment]
Update on GMTNet Paddle reproduction validation:

I have pushed a new commit `36a99ba` to update the Paddle GMTNet implementation, checkpoint, and smoke test.

Final local validation on the full test split:

* tested: 472
* success: 472
* failed: 0
* Paddle Test MAE_sum style: 2.7317654939399936
* Paddle Test Fnorm: 1.7541764742355415

This matches the PyTorch reference result:

* PyTorch Test MAE: about 2.7318
* PyTorch Test Fnorm: about 1.7542

Additional checks completed:

* Layer-by-layer Paddle/PyTorch forward alignment was verified.
* The final Paddle checkpoint has been updated.
* `property_prediction/configs/gmtnet/test.py` now loads the official checkpoint and passes a CPU smoke test.
* The smoke test output shape is `[1, 3, 3]`.

Note:

The full dielectric dataset is not included in this PR. The full validation was performed locally using the available preprocessed data.

**Comment by leeleolay** [comment]
辛苦人工确认

File: property_prediction/configs/gmtnet/paddle_model.pdparams
Comment (leeleolay) [inline property_prediction/configs/gmtnet/paddle_model.pdparams:1]: ？

File: property_prediction/configs/gmtnet/test.py
Comment (leeleolay) [inline property_prediction/configs/gmtnet/test.py:1]: ？

File: property_prediction/configs/gmtnet/train.py
Comment (leeleolay) [inline property_prediction/configs/gmtnet/train.py:1]: 不符合套件风格

File: property_prediction/configs/gmtnet/gmtnet_jarvis_dielectric.yaml
Comment (leeleolay) [inline property_prediction/configs/gmtnet/gmtnet_jarvis_dielectric.yaml:1]: 不符合套件风格

**Comment by cjywahaha-oss** [comment]
The GMTNet integration has been updated in commit 2767499.

The implementation, PropertyPredictor integration, CIF handling, packaging, clean installation, and final CPU regression have been completed. The PR description has also been updated to reflect the final implementation and validation results.

Please review it when convenient. Thank you.

File: setup.py
Comment (leeleolay) [inline setup.py:1]: 为什么修改这个文件

File: MANIFEST.in
Comment (leeleolay) [inline MANIFEST.in:1]: ？

File: property_prediction/predict.py
Comment (leeleolay) [inline property_prediction/predict.py:1]: 改文件是公共使用的，其他模型是否兼容？

File: property_prediction/configs/gmtnet/test.py
Comment (leeleolay) [reply property_prediction/configs/gmtnet/test.py:1]: 为什么新增

File: property_prediction/configs/gmtnet/README.md
Comment (leeleolay) [inline property_prediction/configs/gmtnet/README.md:1]: 不符合套件已有规范

File: property_prediction/configs/gmtnet/paddle_model.pdparams
Comment (leeleolay) [reply property_prediction/configs/gmtnet/paddle_model.pdparams:1]: 不上传权重文件

File: ppmat/models/__init__.py [hunk 1]
```diff
@@ -55,6 +96,9 @@
     "MatterGen",
     "MatterGenWithCondition",
     "DimeNetPlusPlus",
+    "GMTNet",
+    "GMTNetGraphConverter",
+    "GMTNetPredictorInputAdapter",
```
Comment (leeleolay) [inline ppmat/models/__init__.py]: 为什么加一个adapter

File: ppmat/models/__init__.py [hunk 2]
```diff
@@ -131,6 +175,16 @@ def build_graph_converter(cfg: Dict):
...
 
 
+def build_predictor_input_adapter(cfg: Dict, graph_converter):
+    """Build an opt-in structure input adapter for a property predictor."""
+    if cfg is None:
+        return None
+    cfg = copy.deepcopy(cfg)
+    class_name = cfg.pop("__class_name__")
+    init_params = cfg.pop("__init_params__")
+    return eval(class_name)(graph_converter=graph_converter, **init_params)
```
Comment (leeleolay) [inline ppmat/models/__init__.py]: 应该不需要的

File: ppmat/models/__init__.py [hunk 3]
```diff
@@ -35,17 +36,57 @@
...
-from ppmat.models.mateno.mateno import MatENO
 from ppmat.utils import download
 from ppmat.utils import logger
 from ppmat.utils import save_load
 
+
+_OPTIONAL_MODEL_IMPORTS = {
+    "InfGCN": ("ppmat.models.infgcn.infgcn", "InfGCN"),
+    "MatENO": ("ppmat.models.mateno.mateno", "MatENO"),
+}
```
Comment (leeleolay) [inline ppmat/models/__init__.py]: ？

File: ppmat/models/__init__.py [hunk 4]
```diff
@@ -35,17 +36,57 @@
...
+        return _resolve_optional_model_class(class_name)
+    return eval(class_name)
+
+
+def __getattr__(name: str):
+    """Resolve optional model exports only when callers request them."""
+    if name in _OPTIONAL_MODEL_IMPORTS:
+        return _resolve_optional_model_class(name)
+    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
+
```
Comment (leeleolay) [inline ppmat/models/__init__.py]: ？

File: ppmat/models/gmtnet/gmtnet_predictor_adapter.py
Comment (leeleolay) [inline ppmat/models/gmtnet/gmtnet_predictor_adapter.py:1]: 已有的predictor不支持吗？应当在原先的基础上新增

File: ppmat/models/gmtnet/gmtnet_graph_converter.py
Comment (leeleolay) [inline ppmat/models/gmtnet/gmtnet_graph_converter.py:1]: 已有的build graph呢

File: ppmat/datasets/prepare_gmtnet_dataset.py
Comment (leeleolay) [inline ppmat/datasets/prepare_gmtnet_dataset.py:1]: 文件内容是否有点冗余，文件名建议split开头

File: property_prediction/configs/gmtnet/split_gmtnet_dielectric_seed32.json
Comment (leeleolay) [inline property_prediction/configs/gmtnet/split_gmtnet_dielectric_seed32.json:1]: 改文件建议先移动到config/gmtnet下面

File: ppmat/datasets/gmtnet_dataset.py
Comment (leeleolay) [inline ppmat/datasets/gmtnet_dataset.py:1]: 改文件未按照已有套件规范

File: ppmat/datasets/collate_fn.py [hunk 5]
```diff
@@ -88,6 +88,18 @@ def __call__(self, batch: List[Any]) -> Any:
...
+
+    def __init__(self):
+        pass
+
+    def __call__(self, batch):
+        from ppmat.datasets.gmtnet_dataset import gmtnet_dielectric_collate_fn
+
+        return gmtnet_dielectric_collate_fn(batch)
+
+
```
Comment (leeleolay) [inline ppmat/datasets/collate_fn.py]: ？

**Review by leeleolay — COMMENTED** [review]
请认真参考意见进行修改

File: property_prediction/configs/gmtnet/paddle_model.pdparams
Comment (cjywahaha-oss) [reply property_prediction/configs/gmtnet/paddle_model.pdparams:1]: 已删除 PR 中的权重文件。模型验证使用仓库外本地 checkpoint，发行包不再包含权重或数据文件。

Comment (cjywahaha-oss) [reply property_prediction/configs/gmtnet/paddle_model.pdparams:1]: 已从 PR 删除 `paddle_model.pdparams`，不再上传或打包权重文件。

File: property_prediction/configs/gmtnet/test.py
Comment (cjywahaha-oss) [reply property_prediction/configs/gmtnet/test.py:1]: 已删除该独立 `test.py`，不再在配置目录维护额外测试入口；GMTNet 验证已归入套件测试和正式 YAML 使用路径。

Comment (cjywahaha-oss) [reply property_prediction/configs/gmtnet/test.py:1]: 已删除该新增的 `test.py`，不再维护配置目录中的独立测试脚本，验证统一归入套件测试。

File: property_prediction/configs/gmtnet/train.py
Comment (cjywahaha-oss) [reply property_prediction/configs/gmtnet/train.py:1]: 已删除配置目录中的独立训练脚本，训练统一使用现有 `property_prediction/train.py` 套件入口，并通过正式 YAML 配置运行。

File: property_prediction/configs/gmtnet/gmtnet_jarvis_dielectric.yaml
Comment (cjywahaha-oss) [reply property_prediction/configs/gmtnet/gmtnet_jarvis_dielectric.yaml:1]: 已将旧 `config.yaml` 重命名为 `gmtnet_jarvis_dielectric.yaml`，并按现有训练/预测入口整理为正式配置，删除了非套件风格的附属用法。

File: setup.py
Comment (cjywahaha-oss) [reply setup.py:1]: 已移除 GMTNet 专用构建逻辑。现在 setup.py 仅通过标准 package_data 声明安装时所需的 YAML、split 和 README；sdist 资源由 MANIFEST.in 配套控制，权重和数据文件均不会纳入发行包。

File: MANIFEST.in
Comment (cjywahaha-oss) [reply MANIFEST.in:1]: 已将 MANIFEST.in 收敛为最小清单：在保留仓库现有必要资源的基础上，仅新增 GMTNet 的正式 YAML、split 和 README，并明确排除权重、pickle 数据、日志和缓存。

File: property_prediction/predict.py
Comment (cjywahaha-oss) [reply property_prediction/predict.py:1]: 已在原有 Predictor 框架上按 converter capability（`build_prediction_input`、`load_structure_from_cif`）扩展。非 GMTNet 保持原 fallback；普通模型仍在 `no_grad` 下运行，GMTNet 保留其 required forward-grad 行为。

File: property_prediction/configs/gmtnet/README.md
Comment (cjywahaha-oss) [reply property_prediction/configs/gmtnet/README.md:1]: 已按套件的正式配置与公共入口规范重写 README，使用正式 YAML、split 和 CLI 名称；checkpoint 已明确为用户本地路径，不再引用仓库内权重或旧测试脚本。

File: ppmat/models/__init__.py (hunk 1 above)
Comment (cjywahaha-oss) [reply ppmat/models/__init__.py]: 已删除独立 Adapter 和 generic builder。GMTNet 所需输入构建能力已合并到 `GMTNetGraphConverter`，`PropertyPredictor` 在原有框架上按 capability 扩展。

File: ppmat/models/__init__.py (hunk 2 above)
Comment (cjywahaha-oss) [reply ppmat/models/__init__.py]: 已移除该不必要的 generic builder/Adapter 入口，`ppmat.models` 仅保留 GMTNet 及其专用 GraphConverter 的必要导出。

File: ppmat/models/__init__.py (hunk 3 above)
Comment (cjywahaha-oss) [reply ppmat/models/__init__.py]: 已恢复 upstream 的 `InfGCN` 直接 import，并删除 lazy resolver；不再保留用于绕过本地依赖环境的 optional-import 逻辑。

File: ppmat/models/__init__.py (hunk 4 above)
Comment (cjywahaha-oss) [reply ppmat/models/__init__.py]: 已恢复 upstream 的 `MatENO` 直接 import，并移除 `importlib`、lazy resolver 和延迟类解析；没有保留 optional-import workaround。

File: ppmat/models/gmtnet/gmtnet_predictor_adapter.py
Comment (cjywahaha-oss) [reply ppmat/models/gmtnet/gmtnet_predictor_adapter.py:1]: 已删除独立 `gmtnet_predictor_adapter.py`。GMTNet 支持已在原有 `PropertyPredictor` 基础上按 converter capability 增加，必要输入构建能力合并到 `GMTNetGraphConverter`。

File: ppmat/models/gmtnet/gmtnet_graph_converter.py
Comment (cjywahaha-oss) [reply ppmat/models/gmtnet/gmtnet_graph_converter.py:1]: 未改动现有模型的通用 build-graph 流程。GMTNetGraphConverter 仅实现该模型所需的对称约束和图构建契约；当前实现已与基准的 472 个测试样本完成逐项 parity 验证。

File: ppmat/datasets/prepare_gmtnet_dataset.py
Comment (cjywahaha-oss) [reply ppmat/datasets/prepare_gmtnet_dataset.py:1]: 已删除冗余的 `prepare_gmtnet_dataset.py`，替换为 `split_gmtnet_dataset.py`；该工具现在只负责 split 的生成和验证。

File: property_prediction/configs/gmtnet/split_gmtnet_dielectric_seed32.json
Comment (cjywahaha-oss) [reply property_prediction/configs/gmtnet/split_gmtnet_dielectric_seed32.json:1]: 已将 split JSON 完整移动到 `property_prediction/configs/gmtnet/split_gmtnet_dielectric_seed32.json`，内容保持不变。

File: ppmat/datasets/gmtnet_dataset.py
Comment (cjywahaha-oss) [reply ppmat/datasets/gmtnet_dataset.py:1]: 已按现有 Dataset 规范整理属性声明、资源 fallback、Data 字段和 data_index 类型，并删除专用 collate 路径，统一使用 DefaultCollator；正式 split 的内容和顺序保持不变。

File: ppmat/datasets/collate_fn.py (hunk 5 above)
Comment (cjywahaha-oss) [reply ppmat/datasets/collate_fn.py]: 已删除 GMTNet 专用 Collator，train/val/test 三个 loader 均改为使用现有 `DefaultCollator`；batch size 1/2 的输入、预测和 loss 契约已验证。

**Review by leeleolay — COMMENTED** [review]
.
